docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,679 @@
+ */
+typedef int (*enclave_pfhandler_t)(const sgx_pfinfo *pfinfo);
+
+/*
+ * Register the EMM handler with the global exception handler registry
+ * The Runtime should ensure this handler is called first in case of
+ * a #PF before all other handlers.
+ *
+ * @param[in] pfhandler, the EMM page fault handler.
+ * @retval true on success.
+ * @retval false on failure.
+ */
+bool sgx_register_pfhandler(enclave_pfhandler_t pfhandler);
+
+/**
+ * Unregister the EMM handler with the global exception handler registry
+ * @param[in] pfhandler, the EMM page fault handler.
+ * @retval true on success.
+ * @retval false on failure.
+ */
+bool sgx_unregister_pfhandler(enclave_pfhandler_t pfhandler);
+
+```
+
+### OCalls
+
+```
+/*
+ * Call OS to reserve region for EAUG, immediately or on-demand
+ *
+ * @param[in] addr, desired page aligned start address, NULL if no desired address
+ * @param[in] length, size of the region in multiples of page size in bytes
+ * @param[in] flags,  a bitwise OR of flags describing committing mode, committing
+ *                     order, address preference, page type. The untrusted side
+ *    implementation should always invoke mmap syscall with MAP_SHARED|MAP_FIXED_NOREPLACE, and
+ *    translate following additional bits to proper parameters invoking mmap or other SGX specific
+ *    syscall(s) provided by the kernel.
+ *        The flags param of this interface should include exactly one of following for committing mode:
+ *            - EMA_RESERVE: kernel map an address range with PROT_NONE, no EPC EAUGed.
+ *            - EMA_COMMIT_NOW: reserves memory range with EMA_PROT_READ|EMA_PROT_WRITE, if supported, 
+ *                   kernel is given a hint to EAUG EPC pages for the area as soon as possible.
+ *            - EMA_COMMIT_ON_DEMAND: reserves memory range, EPC pages can be EAUGed upon #PF.
+ *        ORed with zero or one of the committing order flags:
+ *            - EMA_GROWSDOWN: if supported, a hint given for the kernel to EAUG pages from higher
+ *                              to lower addresses, no gaps in addresses above the last committed.
+ *            - EMA_GROWSUP: if supported, a hint given for the kernel to EAUG pages from lower
+ *                              to higher addresses, no gaps in addresses below the last committed.
+ *        Optionally ORed with
+ *            -  EMA_ALIGNED(n):	Align the region on a requested	boundary.
+ *                           Fail if a suitable region cannot be found,
+ *                           The argument n specifies the binary logarithm of
+ *                           the desired alignmentand must be at least 12.
+ *        Optionally ORed with one of following page types:
+ *             - EMA_PAGE_TYPE_REG: regular page type.This is the default if not specified.
+ *             - EMA_PAGE_TYPE_SS_FIRST: the first page in shadow stack.
+ *             - EMA_PAGE_TYPE_SS_REST: the rest page in shadow stack. 
+ * @retval 0 on success.
```

Comment:
```suggestion
 * @retval 0 The operation was successful.
```
Doxygen generates a table, so after the return value should be a statement to appear in the 2nd column of the table.  Recommend phrasing it like an error message, as a sentence.